Standalone lines should be removed from the template

As far as I can tell, this implementation doesn't follow [the Mustache spec](https://github.com/mustache/spec/blob/83b0721610a4e11832e83df19c73ace3289972b9/specs/sections.yml#L211-L237) when it comes to trimming lines that only contain tags (and whitespace).

Any reason why you're not testing against the official spec tests?

Hi @kainjow , thanks for the great work on the library. 

I wanted to check the experimental `standalone-lines` branch. 
I do not know if this is because the branch is a work in progress, but while compiling the branch `standalone-lines` in Ubuntu 16.04 I get the following compilation errors:  
~~~
Scanning dependencies of target mustache
[ 50%] Building CXX object CMakeFiles/mustache.dir/tests.cpp.o
In file included from /home/straversaro/src/Mustache/tests.cpp:29:0:
/home/straversaro/src/Mustache/mustache.hpp:498:32: error: declaration of ‘kainjow::mustache::delimiter_set<string_type> kainjow::mustache::context_internal<string_type>::delimiter_set’ [-fpermissive]
     delimiter_set<string_type> delimiter_set;
                                ^
/home/straversaro/src/Mustache/mustache.hpp:399:7: error: changes meaning of ‘delimiter_set’ from ‘class kainjow::mustache::delimiter_set<string_type>’ [-fpermissive]
 class delimiter_set {
       ^
/home/straversaro/src/Mustache/mustache.hpp:524:49: error: declaration of ‘std::shared_ptr<kainjow::mustache::delimiter_set<string_type> > kainjow::mustache::tag<string_type>::delimiter_set’ [-fpermissive]
     std::shared_ptr<delimiter_set<string_type>> delimiter_set;
                                                 ^
/home/straversaro/src/Mustache/mustache.hpp:399:7: error: changes meaning of ‘delimiter_set’ from ‘class kainjow::mustache::delimiter_set<string_type>’ [-fpermissive]
 class delimiter_set {
       ^
/home/straversaro/src/Mustache/mustache.hpp:557:22: error: declaration of ‘kainjow::mustache::tag<string_type> kainjow::mustache::component<string_type>::tag’ [-fpermissive]
     tag<string_type> tag;
                      ^
/home/straversaro/src/Mustache/mustache.hpp:519:7: error: changes meaning of ‘tag’ from ‘class kainjow::mustache::tag<string_type>’ [-fpermissive]
 class tag {
       ^
CMakeFiles/mustache.dir/build.make:62: recipe for target 'CMakeFiles/mustache.dir/tests.cpp.o' failed
make[2]: *** [CMakeFiles/mustache.dir/tests.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/mustache.dir/all' failed
make[1]: *** [CMakeFiles/mustache.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
~~~

--- a/mustache.hpp
+++ b/mustache.hpp
@@ -501,7 +501,7 @@ template <typename string_type>
 class line_buffer_state {
 public:
     string_type data;
-    bool contained_tag = false;
+    bool contained_section_tag = false;
 
     bool is_empty_or_contains_only_whitespace() const {
         for (const auto ch : data) {
@@ -515,7 +515,7 @@ public:
 
     void clear() {
         data.clear();
-        contained_tag = false;
+        contained_section_tag = false;
     }
 };
 
@@ -965,7 +965,7 @@ private:
         // if the line had tags in it, and also if the line is now empty or
         // contains whitespace only. if this situation is true, skip the line.
         bool output = true;
-        if (ctx.line_buffer.contained_tag && ctx.line_buffer.is_empty_or_contains_only_whitespace()) {
+        if (ctx.line_buffer.contained_section_tag && ctx.line_buffer.is_empty_or_contains_only_whitespace()) {
             output = false;
         }
         if (output) {
@@ -977,13 +977,9 @@ private:
         ctx.line_buffer.clear();
     }
     
-    void render_result(const render_handler& handler, context_internal<string_type>& ctx, const string_type& text, bool from_tag) const {
-        (void)ctx;
-        if (from_tag) {
-            //ctx.line_buffer.contained_tag = true;
-        }
-        //ctx.line_buffer.data.append(text);
-        handler(text);
+    void render_result(const render_handler& handler, context_internal<string_type>& ctx, const string_type& text) const {
+        ctx.line_buffer.data.append(text);
+        //handler(text);
         (void)handler;
     }
 
@@ -992,7 +988,7 @@ private:
             if (comp.is_newline()) {
                 render_current_line(handler, ctx, &comp);
             } else {
-                render_result(handler, ctx, comp.text, false);
+                render_result(handler, ctx, comp.text);
             }
             return component<string_type>::walk_control::walk;
         }
@@ -1015,7 +1011,13 @@ private:
                             return component<string_type>::walk_control::stop;
                         }
                     } else if (!var->is_false() && !var->is_empty_list()) {
+                        // account for the section begin tag
+                        ctx.line_buffer.contained_section_tag = true;
+                        
                         render_section(handler, ctx, comp, var);
+                        
+                        // ctx may have been cleared. account for the section end tag
+                        ctx.line_buffer.contained_section_tag = true;
                     }
                 }
                 return component<string_type>::walk_control::skip;
@@ -1098,9 +1100,9 @@ private:
         };
         if (var->is_lambda2()) {
             const basic_renderer<string_type> renderer{render, render2};
-            render_result(handler, ctx, var->lambda2_value()(text, renderer), false);
+            render_result(handler, ctx, var->lambda2_value()(text, renderer));
         } else {
-            render_result(handler, ctx, render(var->lambda_value()(text)), false);
+            render_result(handler, ctx, render(var->lambda_value()(text)));
         }
         return error_message_.empty();
     }
@@ -1108,7 +1110,7 @@ private:
     bool render_variable(const render_handler& handler, const basic_data<string_type>* var, context_internal<string_type>& ctx, bool escaped) {
         if (var->is_string()) {
             const auto& varstr = var->string_value();
-            render_result(handler, ctx, escaped ? escape_(varstr) : varstr, false);
+            render_result(handler, ctx, escaped ? escape_(varstr) : varstr);
         } else if (var->is_lambda()) {
             const render_lambda_escape escape_opt = escaped ? render_lambda_escape::escape : render_lambda_escape::unescape;
             return render_lambda(handler, var, ctx, escape_opt, {}, false);

--- a/mustache.hpp
+++ b/mustache.hpp
@@ -921,10 +921,6 @@ private:
         parser<string_type> parser{input, ctx, root_component_, error_message_};
     }
     
-    void walk(const typename component<string_type>::walk_callback& callback) {
-        root_component_.walk_children(callback);
-    }
-
     string_type render(context_internal<string_type>& ctx) {
         std::basic_ostringstream<typename string_type::value_type> ss;
         render([&ss](const string_type& str) {
@@ -934,7 +930,7 @@ private:
     }
 
     void render(const render_handler& handler, context_internal<string_type>& ctx) {
-        walk([&handler, &ctx, this](component<string_type>& comp) -> typename component<string_type>::walk_control {
+        root_component_.walk_children([&handler, &ctx, this](component<string_type>& comp) -> typename component<string_type>::walk_control {
             return render_component(handler, ctx, comp);
         });
     }

--- a/mustache.hpp
+++ b/mustache.hpp
@@ -498,10 +498,33 @@ private:
 };
 
 template <typename string_type>
+class line_buffer_state {
+public:
+    string_type data;
+    bool contained_tag = false;
+
+    bool is_empty_or_contains_only_whitespace() const {
+        for (const auto ch : data) {
+            // don't look at newlines
+            if (ch != ' ' && ch != '\t') {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    void clear() {
+        data.clear();
+        contained_tag = false;
+    }
+};
+
+template <typename string_type>
 class context_internal {
 public:
     basic_context<string_type>& ctx;
     delimiter_set<string_type> delim_set;
+    line_buffer_state<string_type> line_buffer;
     
     context_internal(basic_context<string_type>& a_ctx)
         : ctx(a_ctx)
@@ -933,11 +956,44 @@ private:
         root_component_.walk_children([&handler, &ctx, this](component<string_type>& comp) -> typename component<string_type>::walk_control {
             return render_component(handler, ctx, comp);
         });
+        // process the last line
+        render_current_line(handler, ctx, nullptr);
+    }
+    
+    void render_current_line(const render_handler& handler, context_internal<string_type>& ctx, const component<string_type>* comp) const {
+        // We're at the end of a line, so check the line buffer state to see
+        // if the line had tags in it, and also if the line is now empty or
+        // contains whitespace only. if this situation is true, skip the line.
+        bool output = true;
+        if (ctx.line_buffer.contained_tag && ctx.line_buffer.is_empty_or_contains_only_whitespace()) {
+            output = false;
+        }
+        if (output) {
+            handler(ctx.line_buffer.data);
+            if (comp) {
+                handler(comp->text);
+            }
+        }
+        ctx.line_buffer.clear();
+    }
+    
+    void render_result(const render_handler& handler, context_internal<string_type>& ctx, const string_type& text, bool from_tag) const {
+        (void)ctx;
+        if (from_tag) {
+            //ctx.line_buffer.contained_tag = true;
+        }
+        //ctx.line_buffer.data.append(text);
+        handler(text);
+        (void)handler;
     }
 
     typename component<string_type>::walk_control render_component(const render_handler& handler, context_internal<string_type>& ctx, component<string_type>& comp) {
         if (comp.is_text()) {
-            handler(comp.text);
+            if (comp.is_newline()) {
+                render_current_line(handler, ctx, &comp);
+            } else {
+                render_result(handler, ctx, comp.text, false);
+            }
             return component<string_type>::walk_control::walk;
         }
         
@@ -970,7 +1026,7 @@ private:
                 return component<string_type>::walk_control::skip;
             case tag_type::partial:
                 if ((var = ctx.ctx.get_partial(tag.name)) != nullptr && (var->is_partial() || var->is_string())) {
-                    const auto partial_result = var->is_partial() ? var->partial_value()() : var->string_value();
+                    const auto& partial_result = var->is_partial() ? var->partial_value()() : var->string_value();
                     basic_mustache tmpl{partial_result};
                     tmpl.set_custom_escape(escape_);
                     if (!tmpl.is_valid()) {
@@ -1042,17 +1098,17 @@ private:
         };
         if (var->is_lambda2()) {
             const basic_renderer<string_type> renderer{render, render2};
-            handler(var->lambda2_value()(text, renderer));
+            render_result(handler, ctx, var->lambda2_value()(text, renderer), false);
         } else {
-            handler(render(var->lambda_value()(text)));
+            render_result(handler, ctx, render(var->lambda_value()(text)), false);
         }
         return error_message_.empty();
     }
     
     bool render_variable(const render_handler& handler, const basic_data<string_type>* var, context_internal<string_type>& ctx, bool escaped) {
         if (var->is_string()) {
-            const auto varstr = var->string_value();
-            handler(escaped ? escape_(varstr) : varstr);
+            const auto& varstr = var->string_value();
+            render_result(handler, ctx, escaped ? escape_(varstr) : varstr, false);
         } else if (var->is_lambda()) {
             const render_lambda_escape escape_opt = escaped ? render_lambda_escape::escape : render_lambda_escape::unescape;
             return render_lambda(handler, var, ctx, escape_opt, {}, false);

--- a/mustache.hpp
+++ b/mustache.hpp
@@ -977,10 +977,8 @@ private:
         ctx.line_buffer.clear();
     }
     
-    void render_result(const render_handler& handler, context_internal<string_type>& ctx, const string_type& text) const {
+    void render_result(context_internal<string_type>& ctx, const string_type& text) const {
         ctx.line_buffer.data.append(text);
-        //handler(text);
-        (void)handler;
     }
 
     typename component<string_type>::walk_control render_component(const render_handler& handler, context_internal<string_type>& ctx, component<string_type>& comp) {
@@ -988,7 +986,7 @@ private:
             if (comp.is_newline()) {
                 render_current_line(handler, ctx, &comp);
             } else {
-                render_result(handler, ctx, comp.text);
+                render_result(ctx, comp.text);
             }
             return component<string_type>::walk_control::walk;
         }
@@ -1100,10 +1098,10 @@ private:
         };
         if (var->is_lambda2()) {
             const basic_renderer<string_type> renderer{render, render2};
-            render_result(handler, ctx, var->lambda2_value()(text, renderer));
+            render_result(ctx, var->lambda2_value()(text, renderer));
         } else {
             render_current_line(handler, ctx, nullptr);
-            render_result(handler, ctx, render(var->lambda_value()(text)));
+            render_result(ctx, render(var->lambda_value()(text)));
         }
         return error_message_.empty();
     }
@@ -1111,7 +1109,7 @@ private:
     bool render_variable(const render_handler& handler, const basic_data<string_type>* var, context_internal<string_type>& ctx, bool escaped) {
         if (var->is_string()) {
             const auto& varstr = var->string_value();
-            render_result(handler, ctx, escaped ? escape_(varstr) : varstr);
+            render_result(ctx, escaped ? escape_(varstr) : varstr);
         } else if (var->is_lambda()) {
             const render_lambda_escape escape_opt = escaped ? render_lambda_escape::escape : render_lambda_escape::unescape;
             return render_lambda(handler, var, ctx, escape_opt, {}, false);

--- a/mustache.hpp
+++ b/mustache.hpp
@@ -1061,7 +1061,6 @@ private:
     };
     
     bool render_lambda(const render_handler& handler, const basic_data<string_type>* var, context_internal<string_type>& ctx, render_lambda_escape escape, const string_type& text, bool parse_with_same_context) {
-        render_current_line(handler, ctx, nullptr);
         const typename basic_renderer<string_type>::type2 render2 = [this, &ctx, parse_with_same_context, escape](const string_type& text, bool escaped) {
             const auto process_template = [this, &ctx, escape, escaped](basic_mustache& tmpl) -> string_type {
                 if (!tmpl.is_valid()) {
@@ -1103,6 +1102,7 @@ private:
             const basic_renderer<string_type> renderer{render, render2};
             render_result(handler, ctx, var->lambda2_value()(text, renderer));
         } else {
+            render_current_line(handler, ctx, nullptr);
             render_result(handler, ctx, render(var->lambda_value()(text)));
         }
         return error_message_.empty();

--- a/mustache.hpp
+++ b/mustache.hpp
@@ -1061,6 +1061,7 @@ private:
     };
     
     bool render_lambda(const render_handler& handler, const basic_data<string_type>* var, context_internal<string_type>& ctx, render_lambda_escape escape, const string_type& text, bool parse_with_same_context) {
+        render_current_line(handler, ctx, nullptr);
         const typename basic_renderer<string_type>::type2 render2 = [this, &ctx, parse_with_same_context, escape](const string_type& text, bool escaped) {
             const auto process_template = [this, &ctx, escape, escaped](basic_mustache& tmpl) -> string_type {
                 if (!tmpl.is_valid()) {

--- a/tests.cpp
+++ b/tests.cpp
@@ -1317,6 +1317,24 @@ TEST_CASE("standalone_lines") {
             "| A Line"
         );
     }
-    
+
+    SECTION("crlf") {
+        mustache tmpl{"|\r\n{{#boolean}}\r\n{{/boolean}}\r\n|"};
+        data data("boolean", true);
+        CHECK(tmpl.render(data) == "|\r\n|");
+    }
+
+    SECTION("without_previous_line") {
+        mustache tmpl{"  {{#boolean}}\n#{{/boolean}}\n/"};
+        data data("boolean", true);
+        CHECK(tmpl.render(data) == "#\n/");
+    }
+
+    SECTION("without_next_newline") {
+        mustache tmpl{"#{{#boolean}}\n/\n  {{/boolean}}"};
+        data data("boolean", true);
+        CHECK(tmpl.render(data) == "#\n/\n");
+    }
+
 }
 

--- a/tests.cpp
+++ b/tests.cpp
@@ -1282,5 +1282,41 @@ TEST_CASE("standalone_lines") {
         CHECK(root_children[14].children.empty());
     }
     
+    SECTION("remove_standalone_lines") {
+        mustache tmpl{
+            "|\n"
+            "| This Is\n"
+            "{{#boolean}}\n"
+            "|\n"
+            "{{/boolean}}\n"
+            "| A Line"
+        };
+        data data("boolean", true);
+        CHECK(tmpl.render(data) ==
+            "|\n"
+            "| This Is\n"
+            "|\n"
+            "| A Line"
+        );
+    }
+    
+    SECTION("remove_indented_standalone_lines") {
+        mustache tmpl{
+            "|\n"
+            "| This Is\n"
+            "  {{#boolean}}\n"
+            "|\n"
+            "  {{/boolean}}\n"
+            "| A Line"
+        };
+        data data("boolean", true);
+        CHECK(tmpl.render(data) ==
+            "|\n"
+            "| This Is\n"
+            "|\n"
+            "| A Line"
+        );
+    }
+    
 }
 

